Add support for markdown preview in compose area.

hackerkid opened the pull request:
Under development

- [X] Preview using echo.apply_markdown
- [x] CasperJs Frontend Tests

Fixes https://github.com/zulip/zulip/issues/217

hackerkid commented:
@timabbott A basic version is completed. can you check and give a feedback when you are free.

showell commented:
Looks good so far.  Tim's out this week, but if you can finish up your todos, we should be able to merge this early next week.  You might want to go the "frontend" thread on https://zulip.tabbott.net/# and ask about the preview button.

hackerkid commented:
@showell Sure :+1:

hackerkid commented:
Here are some screenshots. 

![screenshot from 2016-09-01 16 43 31](https://cloud.githubusercontent.com/assets/7190633/18165806/00cfe848-7065-11e6-8290-3deafcfef5bd.png)
![screenshot from 2016-09-01 16 43 41](https://cloud.githubusercontent.com/assets/7190633/18165804/009e5b84-7065-11e6-971c-2b9a9a4f9b8c.png)
![screenshot from 2016-09-01 17 00 04](https://cloud.githubusercontent.com/assets/7190633/18165898/8d0d2e24-7065-11e6-8850-1e323f712c61.png)

hackerkid commented:
The icons have been replaced with text for `preview` and `write` modes. 

![screenshot from 2016-09-06 01 42 47](https://cloud.githubusercontent.com/assets/7190633/18256460/4c884674-73d3-11e6-8068-5c203ed12a0d.png)

![screenshot from 2016-09-06 01 42 57](https://cloud.githubusercontent.com/assets/7190633/18256459/4c802ef8-73d3-11e6-97fe-730e805cb5a6.png)

timabbott commented:
Looks like this has some lint failures, and I imagine we'll eventually want to tweak the visual design for the Preview/Write buttons, but this looks pretty good.  One question: what's the story with the scrollbar on the compose box that shows up here?  I thikn that's a regression with this PR...

timabbott commented on static/styles/zulip.css:
```diff
@@ -3547,3 +3547,15 @@ li.show-more-private-messages a {
 
+#preview_message_area {
+    max-height: 22em;
+    overflow: scroll;
```
should this be `overflow:auto` perhaps?

timabbott commented on frontend_tests/casper_tests/04-compose.js:
```diff
@@ -117,6 +117,26 @@ casper.waitUntilVisible('#compose', function () {
+casper.waitUntilVisible('#markdown_preview', function () {
+    casper.test.assertNotVisible('#undo_markdown_preview', 'Write button is hidden');
+    casper.click("#markdown_preview");
+    casper.test.assertNotVisible('#markdown_preview', 'Preview button is hidden.');
```
I think you might want to do this as a waitWhileVisible (and similarly with the test below) so the browser has time to rerender the page; the current code looks like it'll potentially have nondeterministic errors...

timabbott commented:
@hackerkid I see you pushed a new version -- does that version address my comments above?

hackerkid commented:
@timabbott Almost.  I have fixed everything you suggested except adding `waitWhileVisible` for the second test (I am not able to pass that test). I have also made some CSS changes. While working on these I found some new bugs (like the preview still shows up when we close and reopen the compose message). Working on fixing all these bugs.

hackerkid commented:
@timabbott Fixed all known issues. Please check.

showell commented on templates/zerver/compose.html:
```diff
@@ -78,6 +79,9 @@
                      id="attach_files" href="#" title="{{ _('Attach files') }}"></a>
                   <a class="message-control-button icon-vector-font"
                      href="#markdown-help" title="Formatting" data-toggle="modal"></a>
+                  <a class="preview_text" style="display:none;"
```
I don't think the name `preview_text` is very intuitive.  You don't need to "preview" text; it's already text.

hackerkid replied:
@showell Changed the name to `markdown_preview_button` as we have to replace the `text` with a suitable icon in the near future.

showell commented on templates/zerver/compose.html:
```diff
@@ -78,6 +79,9 @@
                      id="attach_files" href="#" title="{{ _('Attach files') }}"></a>
                   <a class="message-control-button icon-vector-font"
                      href="#markdown-help" title="Formatting" data-toggle="modal"></a>
+                  <a class="markdown_preview_button" style="display:none;"
```
I'm sorry, I still find this confusing.  You now have a tag whose class is markdown_preview_button but its id is undo_markdown_preview.  Somebody scanning the code is going to think this is the button that turns on preview when they read the class, which comes first.  I know the class refers to overall feature, but I'm not sure you really need it.  In the CSS where the two buttons have characteristics in common, you can just use both id selectors separated by a comma.

hackerkid replied:
@showell I have removed the extra class. Please check.

showell commented on static/styles/zulip.css:
```diff
@@ -3547,3 +3547,16 @@ li.show-more-private-messages a {
+    min-height: 58px;
+}
+
+a#markdown_preview, a#undo_markdown_preview {
```
Thanks for doing this!  Our convention is one selector per line, like this:

```
a#markdown_preview,
a#undo_markdown_preview {
```

hackerkid replied:
@showell Fixed :)

timabbott commented on frontend_tests/casper_tests/04-compose.js:
```diff
@@ -141,6 +141,45 @@ casper.then(function () {
+            });
+        });
+    });
+});
```
@umairwaheed this sort of nesting of Casper test elements doesn't work reliably, right?

umairwaheed replied:
I am not too much sure about nesting, the documentation only says that to avoid race conditions `waitFor*` functions shouldn't be chained together. `casper.waitUntilVisible('#undo_markdown_preview')` should be wrapped in a `then` otherwise it will have a race condition with `casper.waitUntilVisible('#markdown_preview'`. Of course, race condition will only be a problem if both tests are contending for the same shared resources.

timabbott commented:
Just played around with this -- This looks pretty good!  There are a few things I think we could improve:
* The Casper test issue above I'm home Umair will comment on.  I think this is the main blocker for merging this.
* I think we could do some tweaks to how the rendered preview is styled (e..g I might show the Sender name and increase the top margin/padding to make it look a bit more similar to already-sent messages).  If you want help with this, I imagine @brockwhittaker will have some ideas.  I'm OK with merging an imperfect version and planning to tweak it after.
* In cases where the backend Bugdown renderer doesn't agree with the frontend rendered, I'd like to actually render the message using the backend and replace the rendering based on the result if it produces different HTML.  You can do the rendering with the `GET messages/render` endpoint.  I don't think this is a blocker for merging this, but it's something we should work on soon in order to complete this feature.

brockwhittaker commented:
Hey @hackerkid! The preview looks/works pretty well. I added some styling to make it more uniform to the textarea. I also changed the text to icons. I think from a design perspective adding the icons you had originally looks better (more uniform), however that's of course a reasonable subject for debate.

Check out the changes here: https://github.com/brockwhittaker/zulip/commit/f0d9bce0e183562597f8d0b158745b5df7253239

showell commented:
No big debate here... I want to get this change merged and we can iterate as needed.

hackerkid commented:
@brockwhittaker Thanks for the changes. I have updated the PR to include your changes. @umairwaheed @timabbott I have removed the nesting. Please check.

timabbott commented:
Looks great!  I merged it, after cleaning up the title/href values for the buttons.

@hackerkid I think we should soon implement rendering with the backend markdown parser and swapping in the result if different, so that we don't give the wrong idea; do you want to take a crack at that?

hackerkid commented:
@timabbott Yeah Sure.

brainwane commented:
@hackerkid @timabbott @brockwhittaker  A Zulip user just called the preview button "life-changing" and said "Thanks Zulip team!" Passing it along. :)

jjst commented:
The preview button is super useful! Thanks <3

showell commented:
Congrats!
